sematic/caching/caching.py

```diff
@@ -0,0 +1,144 @@
+# Standard Library
+import json
+import logging
+from typing import Callable, Optional, Union
+
+# Sematic
+from sematic.abstract_future import AbstractFuture
+from sematic.types.serialization import (
+    get_json_encodable_summary,
+    type_to_json_encodable,
+    value_to_json_encodable,
+)
+from sematic.utils.hashing import get_str_sha1_digest, get_value_and_type_sha1_digest
+
+CACHE_NAMESPACE_MAX_LENGTH = 50
+
+CacheNamespaceCallable = Callable[[AbstractFuture], str]
+CacheNamespace = Optional[Union[str, CacheNamespaceCallable]]
+
+logger = logging.getLogger(__name__)
+
+
+def resolve_cache_namespace(
+    cache_namespace: CacheNamespace, root_future: AbstractFuture
+) -> str:
+    """
+    Returns a string cache namespace from a `CacheNamespace`, executing the `Callable`
+    instances of this type, if it is the case.
+
+    The string representation is truncated to 50 characters, in order to be a
+    human-readable part of the final cache key for each `Future` execution.
+
+    Parameters
+    ----------
+    cache_namespace: CacheNamespace
+        A string or a `Callable` which returns a string, which is used as the cache key
+        namespace.
+    root_future: AbstractFuture
+        The root Future of the pipeline for which to look up the code git commit status.
+
+    Returns
+    -------
+    A string of maximum length 50.
+    """
```

Comment:
It might be helpful to raise an error if the given future is not the root. If it is the root, it should have `parent_future == None`

Reply:
Done. Introduced an `is_root_future` on `Future` and propagated it some. Some propagations triggered linting errors that I was not willing to inhibit, so didn't propagate everywhere.

Also added test cases.